security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -0,0 +1,195 @@
+	ret := &SecretCache{
+		caClient:         cl,
+		rotationInterval: rotationInterval,
+		evictionDuration: evictionDuration,
+		secretTTL:        secretTTL,
+		closing:          make(chan bool, 1),
+	}
+
+	atomic.StoreUint64(&ret.secretChangedCount, 0)
+	go ret.keyRotationJob()
+	return ret
+}
+
+// GetSecret gets secret from cache.
+func (sc *SecretCache) GetSecret(proxyID, jwtToken string) (*SecretItem, error) {
+	now := time.Now()
+
+	// Return directly if secret could be found from cache and not expired.
+	val, found := sc.secrets.Load(proxyID)
+	if found {
+		s := val.(SecretItem)
+		if s.jwtToken == jwtToken && !sc.checkExpired(&s) {
+			// Update cached secret's lastUsedTime.
+			s.lastUsedTime = now
+			return &s, nil
+		}
+	}
+
+	ns, err := sc.generateSecret(jwtToken, now)
+	if err != nil {
+		log.Errorf("Failed to generate secret for proxy %q: %v", proxyID, err)
+		return nil, err
+	}
+
+	sc.secrets.Store(proxyID, *ns)
+	return ns, nil
+}
+
+// Close shuts down the secret cache.
+func (sc *SecretCache) Close() {
+	sc.closing <- true
+}
+
+func (sc *SecretCache) keyRotationJob() {
+	// Wake up once in a while and refresh stale items.
+	sc.rotationTicker = time.NewTicker(sc.rotationInterval)
+	for {
+		select {
+		case now := <-sc.rotationTicker.C:
+			sc.rotate(now)
+		case <-sc.closing:
+			sc.rotationTicker.Stop()
+			return
+		}
+	}
+}
+
+func (sc *SecretCache) rotate(t time.Time) {
+	sc.secrets.Range(func(key interface{}, value interface{}) bool {
+		proxyID := key.(string)
+		now := time.Now()
+
+		e := value.(SecretItem)
+
+		// Remove from cache if the secret hasn't been used for a while, this prevent the cache growing indefinitely.
+		if now.After(e.lastUsedTime.Add(sc.evictionDuration)) {
+			sc.secrets.Delete(proxyID)
+			return true
+		}
+
+		// Re-generate secret if it's expired.
+		if sc.checkExpired(&e) {
+			go func() {
+				ns, err := sc.generateSecret(e.jwtToken, now)
+				if err != nil {
+					log.Errorf("Failed to generate secret for proxy %q: %v", proxyID, err)
+					return
+				}
+
+				sc.secrets.Store(proxyID, *ns)
+
+				//TODO(quanlin): push new secret to proxy.
+
+				atomic.AddUint64(&sc.secretChangedCount, 1)
+			}()
+		}
+
+		return true
+	})
+}
+
+func (sc *SecretCache) generateSecret(jwtToken string, t time.Time) (*SecretItem, error) {
+	options := util.CertOptions{
+		Host:       jwtToken,
+		RSAKeySize: keySize,
+	}
+
+	// Generate the cert/key, send CSR to CA.
+	csrPEM, keyPEM, err := util.GenCSR(options)
+	if err != nil {
+		return nil, err
+	}
+
+	certChainPER, err := sc.caClient.Sign(csrPEM, jwtToken, int64(sc.secretTTL.Seconds()))
+	if err != nil {
+		return nil, err
+	}
+
+	return &SecretItem{
+		CertificateChain: certChainPER,
+		PrivateKey:       keyPEM,
+		jwtToken:         jwtToken,
+		lastUsedTime:     t,
+		createdTime:      t,
+	}, nil
+}
+
+func (sc *SecretCache) checkExpired(s *SecretItem) bool {
```

Comment:
I think there're some complexities in rotating.

- First, if node agent needs to proactively rotate key cert, it should refresh key cert before TTL reaches, instead of `now.After(...)`.
- Node agent is not always able to send CSR to Google CA and push the results to envoy. If the metadata credential expires already (which is only updated when a pull SDS request comes from envoy), node agent can't impersonate workload.
- Envoy can configure the SDS request frequency like every x seconds. We can make that frequency smaller than the cert 
- So to make node agent proactive refreshing meaningful, you need to make OAuth2 token valid between two SDS request and the key cert TTL is shorter than the SDS request interval.

In summary, for short term we can rely on Envoy's configuration. 

@JimmyCYJ  @mandarjog Can you comment on this?

Reply:
Key rotation here is only for SDS.streamSecret. proxy periodically calls SDS.fetchSecret with interval set in envoy configuration; while for streamSecret, client only calls it once to setup connection, all later update is pushed from SDS to envoy, proactive push is one advantage of envoy v2 api to support grpc.

@qiwzhang, @mangchiandjjoe

Reply:
per decision from last Friday 
updated on server side - 
1. re-generate key/cert for every fetchsecret call.
2. keep current cache/key rotation for streamsecret.

client side(@JimmyCYJ)-
proxy uses fetch instead of stream for now, we could tweak this part later once having better idea of how often token expire, etc.